packages/storage/amplify_storage_s3_dart/lib/src/model/s3_list_result.dart

```diff
@@ -0,0 +1,124 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:amplify_core/amplify_core.dart' hide PaginatedResult;
+import 'package:amplify_storage_s3_dart/amplify_storage_s3_dart.dart';
+import 'package:amplify_storage_s3_dart/src/sdk/s3.dart' as s3;
+import 'package:meta/meta.dart';
+import 'package:smithy/smithy.dart';
+
+/// {@template storage.amplify_storage_s3.list_result}
+/// The result returned by the Storage S3 plugin `list` API.
+/// {@endtemplate}
+class S3ListResult extends StorageListResult<List<S3Item>> {
+  /// {@macro storage.amplify_storage_s3.list_result}
+  S3ListResult(
+    super.items, {
+    required super.hasNextPage,
+    required this.metadata,
+    super.nextToken,
+  });
+
+  /// Creates a [S3ListResult] from a [PaginatedResult] provided by
+  /// smithy. This named constructor should be used internally only.
+  @internal
+  factory S3ListResult.fromPaginatedResult(
+    PaginatedResult<s3.ListObjectsV2Output, int, String> paginatedResult, {
+    required String prefixToDrop,
+  }) {
+    final output = paginatedResult.items;
+    final metadata = S3ListMetadata.fromS3CommonPrefixes(
+      prefixToDrop: prefixToDrop,
+      commonPrefixes: output.commonPrefixes?.toList(),
+      delimiter: output.delimiter,
+    );
+    final items = output.contents
+            ?.map(
+              (s3.S3Object item) => S3Item.fromS3Object(
+                item,
+                prefixToDrop: prefixToDrop,
+              ),
+            )
+            .toList() ??
+        const <S3Item>[];
+
+    return S3ListResult(
+      items,
+      hasNextPage: paginatedResult.hasNext,
+      nextToken: paginatedResult.nextContinuationToken,
+      metadata: metadata,
+    );
+  }
+
+  /// Merges two instances of [S3ListResult] into one.
+  S3ListResult merge(S3ListResult other) {
+    final items = <S3Item>[...this.items, ...other.items];
+    final metadata = this.metadata.merge(other.metadata);
+    return S3ListResult(
+      items,
+      hasNextPage: other.hasNextPage,
+      nextToken: other.nextToken,
+      metadata: metadata,
+    );
+  }
+
+  /// Metadata that is specific to the plugin
+  final S3ListMetadata metadata;
+}
+
+/// The metadata returned from the Storage S3 plugin `list` API.
+class S3ListMetadata {
+  /// Creates a S3ListMetadata from the `commonPrefix` and `delimiter`
+  /// properties of the [s3.ListObjectsV2Output].
+  factory S3ListMetadata.fromS3CommonPrefixes({
+    required String prefixToDrop,
+    List<s3.CommonPrefix>? commonPrefixes,
+    String? delimiter,
+  }) {
+    final extractedSubPath = <String>[];
+
+    if (commonPrefixes != null) {
+      for (final commonPrefix in commonPrefixes) {
+        final prefix = commonPrefix.prefix;
+        if (prefix != null) {
+          extractedSubPath.add(prefix.replaceRange(0, prefixToDrop.length, ''));
```

Comment:
Let's reuse this method if it's meant to be reused - or make more general

```suggestion
          extractedSubPath.add(S3Item.dropPrefixFromKey(prefixToDrop: prefixToDrop, key: prefix));
```